ihm/reader.py

```diff
@@ -2716,6 +2762,102 @@ def __call__(self, catname, keyname, line):
+        mssc.end_state = self.sysr.states.get_by_id_or_none(end_state_id)
+        mssc.dataset_group = \
+            self.sysr.dataset_groups.get_by_id_or_none(dataset_group_id)
+        mssc.details = details
+        # Get the MultiStateScheme
+        mss = self.sysr.multi_state_schemes.get_by_id(scheme_id)
+        # Add the connectivity to the scheme
+        mss.add_connectivity(mssc)
+
+
+class _KineticRateHandler(Handler):
+    category = '_ihm_kinetic_rate'
+
+    def __call__(self, id,
+                 transition_rate_constant,
+                 equilibrium_constant,
+                 equilibrium_constant_determination_method,
+                 equilibrium_constant_unit,
+                 details,
+                 scheme_connectivity_id,
+                 dataset_group_id,
+                 external_file_id):
+        # Get the object or create the object
+        k = self.sysr.kinetic_rates.get_by_id(id)
+        # Add the content
+        k.transition_rate_constant = transition_rate_constant
+        k.equilibrium_constant = equilibrium_constant
+        k.equilibrium_constant_determination_method = \
+            equilibrium_constant_determination_method
+        k.equilibrium_constant_unit = equilibrium_constant_unit
+        k.details = details
+        k.dataset_group = \
+            self.sysr.dataset_groups.get_by_id_or_none(dataset_group_id)
+        k.external_file = \
+            self.sysr.external_files.get_by_id_or_none(external_file_id)
+        tmp_connectivities = self.sysr.multi_state_scheme_connectivities
+        mssc = tmp_connectivities.get_by_id(scheme_connectivity_id)
+        # Add the kinetic rate to the connectivity
+        mssc.kinetic_rate = k
+
+
+class _RelaxationTimeHandler(Handler):
+    category = '_ihm_relaxation_time'
+
+    def __call__(self, id, value, unit, amplitude,
+                 dataset_group_id, external_file_id, details):
+        # Get the object or create the object
+        r = self.sysr.relaxation_times.get_by_id(id)
+        # Add the content
+        r.value = value
+        r.unit = unit
+        r.amplitude = amplitude
+        r.dataset_group = \
+            self.sysr.dataset_groups.get_by_id_or_none(dataset_group_id)
+        r.external_file = \
+            self.sysr.external_files.get_by_id_or_none(external_file_id)
+        r.details = details
+
+
+class _RelaxationTimeMultiStateSchemeHandler(Handler):
+    category = '_ihm_relaxation_time_multi_state_scheme'
+
+    def __call__(self, id, relaxation_time_id,
+                 scheme_id, scheme_connectivity_id,
+                 details):
+        r = self.sysr.relaxation_times.get_by_id(relaxation_time_id)
+        mss = self.sysr.multi_state_schemes.get_by_id(scheme_id)
+        tmp_connectivities = self.sysr.multi_state_scheme_connectivities
+        mssc = tmp_connectivities.get_by_id_or_none(scheme_connectivity_id)
+        # If the relaxation time is assigned to a connectivity, add it there
```

Comment:
What happens if this table is encountered in the file before `_ihm_multi_state_scheme_connectivity`? You may need to store this data temporarily in the class and then assign it to schemes in the `finalize` method. There are several existing tests in `test_reader.py` which read two tables A and B both as A+B and B+A to make sure we get the same results either way. In practice users generate mmCIF files with tables in random order.